feat: add a valueFrom field to Settings

sylr opened the pull request.

sylr committed:
feat: add a valueFrom field to Settings

Signed-off-by: Sylvain Rabot <sylvain@formance.com>

gfyrag commented on internal/resources/settings/helpers.go:
```diff
@@ -393,13 +399,109 @@ func findMatchingSettings(settings []v1beta1.Settings, flattenKeys ...string) (*
+			if setting.Spec.Value != "" {
+				return &setting.Spec.Value, nil
+			}
+			if setting.Spec.ValueFrom != nil {
```
Useless if? 
The oneof should validate we have ether Value or ValueFrom?

gfyrag commented on internal/resources/settings/helpers.go:
```diff
@@ -393,13 +399,109 @@ func findMatchingSettings(settings []v1beta1.Settings, flattenKeys ...string) (*
+				if err != nil {
+					if errors.Is(err, ErrResourceNotFound) || errors.Is(err, ErrOptionalResourceNotFound) {
+						// Try fallback to formance-system namespace
+						value, err2 := resolveValueFrom(ctx, "formance-system", setting.Spec.ValueFrom)
```
Maybe you can get the current namespace of the running pod? By injecting it using the downward api.

flemzord committed:
fix: correct valueFrom fallback logic for optional resources

- Fix bug where optional resource check used wrong error variable (err vs err2)
- Simplify code since CEL validation ensures exactly one of value/valueFrom is set
- Remove test case for impossible state (both value and valueFrom set)
